### ISSUE TITLE
Report warning when program behavior depends on dictionary or set iteration order

### ISSUE THREAD

**Issue by yugr** [PyCQA/pylint#2467]
### Is your feature request related to a problem? Please describe

A frequent bug which I see people running into is forgetting about undefined iteration order over dictionary/set elements:
```
d = { ... }
for x in d:
    print(x)  # Prints elements in arbitrary order
```
This makes program output unstable which in practice causes all sorts of inconveniences and bugs.

It would be great it if Pylint could detect this.

### Describe the solution you'd like

Pylint should emit a warning when program behavior depends on dictionary/set iteration order.

### Additional context

I've attached a draft but working patch. For the following code
```
d = {1:2}
for k in d.keys():
    print(k)
```
it reports
```
tmp.py:3:9: E0130: Program depends on undefined iteration order (undef-iteration-order)
```
I'd be happy to improve it according maintainer's suggestions. 

[unsorted-iteration-0.patch.txt](https://github.com/PyCQA/pylint/files/2343719/unsorted-iteration-0.patch.txt)

**Comment by PCManticore** [comment]
I'm not sure this is such a big of a problem, especially since we can't really figure out the intent of the author of the code. This also looks like it could be a source of frustration as it would emit an error for a relative common pattern.

**Comment by yugr** [comment]
> especially since we can't really figure out the intent of the author of the code

Note that I limit warning to cases when loop performs IO so iteration order is _guaranteed_ to change program semantics.

**Comment by yugr** [comment]
> I'm not sure this is such a big of a problem

Cited as no. 5 in [The 10 Most Common Mistakes That Python Developers Make](https://www.toptal.com/python/top-10-mistakes-that-python-programmers-make) which is the first Google link for "common python bugs".

**Comment by PCManticore** [comment]
@yugr There's no way to figure out that the block of code is doing I/O, unless you plan on having heuristics for every potential Python function that might do I/O. Nonetheless, if you still need this, you can easily write your own checker which you can load with `--load-plugins=`, no need to have it in core pylint per se.

**Comment by yugr** [comment]
> Nonetheless, if you still need this, you can easily write your own checker
> which you can load with --load-plugins=, no need to have it in core pylint per se.

Well, I'd like to hope that my code is free of this sort of bug. My main concern is that the code I see in OSS and in the companies I work at suggests that this bug is quite widespread.

> There's no way to figure out that the block of code is doing I/O,
> unless you plan on having heuristics for every potential Python function that might do I/O

That's a valid concern (although I mainly seen this happening with `print` and `file.write`).

I see your point here but keep in mind that prohibiting heuristic checkers would significantly limit the spectre of bugs Pylint is capable of detecting.
